File: packages/algo-py/solutions/algo_01.py
import queue

DIRECTIONS = [
    [-1, -1],
    [-1, 0],
    [-1, 1],
    [0, -1],
    [0, 1],
    [1, -1],
    [1, 0],
    [1, 1],
]


def minesweeper(board, click):
    clickX, clickY = click[0], click[1]
    if board[clickX][clickY] == "M":
        board[clickX][clickY] = "X"
        return board

    board[clickX][clickY] = "B"
    m, n = len(board), len(board[0])
    q = queue.Queue()
    q.put(clickX*n + clickY)

    while not q.empty():
        point = q.get()
        px, py = point // n, point % n
        mines = __get_adjacent_mines(board, px, py, m, n)
        if mines > 0:
            board[px][py] = str(mines)
        else:
            for direction in DIRECTIONS:
                i, j = px + direction[0], py + direction[1]
                if __is_valid_index(i, j, m, n) and board[i][j] == "E":
                    board[i][j] = "B"
                    q.put(i * n + j)
    return board


def minesweeper_dfs(board, click):
    clickX, clickY = click[0], click[1]
    m, n = len(board), len(board[0])
    if board[clickX][clickY] == "M":
        board[clickX][clickY] = "X"
    else:
        __dfs(board, clickX, clickY, m, n)
    return board


def __dfs(board, i, j, m, n):
    if not __is_valid_index(i, j, m, n) or board[i][j] != "E":
        return

    mines = __get_adjacent_mines(board, i, j, m, n)
    if mines > 0:
        board[i][j] = mines.toString()
    else:
        board[i][j] = "B"
        for direction in DIRECTIONS:
            __dfs(board, i + direction[0], j + direction[1], m, n)


def __get_adjacent_mines(board, x, y, m, n):
    count = 0
    for direction in DIRECTIONS:
        i, j = x + direction[0], y + direction[1]
        if __is_valid_index(i, j, m, n) and board[i][j] == "M":
            count += 1
    return count


def __is_valid_index(i, j, m, n):
    return 0 <= i and i < m and 0 <= j and j < n

File: packages/algo-ts/src/algo_01.ts
import Queue from "./collections/queue";

const DIRECTIONS = [
  [-1, -1],
  [-1, 0],
  [-1, 1],
  [0, -1],
  [0, 1],
  [1, -1],
  [1, 0],
  [1, 1],
];

export function minesweeper(board: string[][], click: number[]): string[][] {
  const [clickX, clickY] = click;
  if (board[clickX][clickY] === "M") {
    board[clickX][clickY] = "X";
    return board;
  }

  board[clickX][clickY] = "B";
  const m = board.length;
  const n = board[0].length;
  const queue = new Queue<number>();
  queue.enqueue(clickX * n + clickY);

  while (!queue.isEmpty()) {
    const point = queue.dequeue()!;
    const px = (point / n) >> 0; // >> will only take the integral part
    const py = point % n;
    const mines = getAdjacentMines(board, px, py, m, n);
    if (mines > 0) {
      board[px][py] = mines.toString();
    } else {
      let i = 0;
      let j = 0;
      for (const direction of DIRECTIONS) {
        i = px + direction[0];
        j = py + direction[1];
        if (isValidIndex(i, j, m, n) && board[i][j] === "E") {
          board[i][j] = "B";
          queue.enqueue(i * n + j);
        }
      }
    }
  }

  return board;
}

export function minesweeperDFS(board: string[][], click: number[]): string[][] {
  const [clickX, clickY] = click;
  const m = board.length;
  const n = board[0].length;
  if (board[clickX][clickY] === "M") {
    board[clickX][clickY] = "X";
  } else {
    dfs(board, clickX, clickY, m, n);
  }
  return board;
}

function dfs(board: string[][], i: number, j: number, m: number, n: number) {
  if (!isValidIndex(i, j, m, n) || board[i][j] != "E") {
    return;
  }

  const mines = getAdjacentMines(board, i, j, m, n);
  if (mines > 0) {
    board[i][j] = mines.toString();
  } else {
    board[i][j] = "B";
    for (const direction of DIRECTIONS) {
      dfs(board, i + direction[0], j + direction[1], m, n);
    }
  }
}

function getAdjacentMines(
  board: string[][],
  x: number,
  y: number,
  m: number,
  n: number
): number {
  let count = 0;
  let i = 0;
  let j = 0;
  for (const direction of DIRECTIONS) {
    i = x + direction[0];
    j = y + direction[1];
    if (isValidIndex(i, j, m, n) && board[i][j] == "M") {
      count++;
    }
  }
  return count;
}

function isValidIndex(i: number, j: number, m: number, n: number): boolean {
  return 0 <= i && i < m && 0 <= j && j < n;
}
